client/src/main/java/org/eclipse/hono/client/impl/RegistrationClientImpl.java

```diff
@@ -171,8 +192,36 @@ public void get(final String deviceId, final Handler<AsyncResult<RegistrationRes
      */
     @Override
     public void assertRegistration(final String deviceId, final Handler<AsyncResult<RegistrationResult>> resultHandler) {
+
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(resultHandler);
-        createAndSendRequest(ACTION_ASSERT, createDeviceIdProperties(deviceId), null, resultHandler);
+
+        final String key = "assert-" + deviceId;
+        getCachedRegistrationAssertion(key).recover(t -> {
+            final Future<RegistrationResult> regResult = Future.future();
+            createAndSendRequest(ACTION_ASSERT, createDeviceIdProperties(deviceId), null, regResult.completer());
+            return regResult.map(response -> {
+                switch(response.getStatus()) {
+                case HttpURLConnection.HTTP_OK:
+                    if (isCachingEnabled()) {
+                        final String token = response.getPayload().getString(RegistrationConstants.FIELD_ASSERTION);
+                        putResponseToCache(key, response, JwtHelper.getExpiration(token).toInstant());
+                    }
+                default:
+                    return response;
+                }
+            });
+        }).setHandler(resultHandler);
+
+    }
+
+    private Future<RegistrationResult> getCachedRegistrationAssertion(final String key) {
+
+        final RegistrationResult result = getResponseFromCache(key);
+        if (result == null) {
+            return Future.failedFuture("cache miss");
```

Comment:
"cache miss or disabled" ?

Reply:
That is a very good question. Currently, we only cache a positive response, i.e. a response containing an assertion. However, we could also cache a negative response in order to prevent round trips to the service, e.g. for 2 minutes. But FMPOV we shouldn't do that and instead always try to get a new assertion when the cache doesn't contain a (positive) response already. WDYT?

Reply:
I find caching "negative answers" considerable, I have to admit, but we can do that later whenever we think (or found out) that it would help us. So don't do it now FMPOV.

However, my comment was meant much simpler: I just proposed to change the String of the failed future, since it would fail not only on cache miss but also if the cache was disabled by configuration.
Is not a functional change of course, more like a documenting improvement (since the String does not play any further role if I see things correctly).

Reply:
> However, my comment was meant much simpler

I know, but it made me realize that we might have an issue here :-)
As you already pointed out the string itself has no real meaning but simply indicates that there was no result in the cache without making any assumptions as to why that is.